pkg/referrerstore/oras/cache.go

```diff
@@ -0,0 +1,107 @@
+package oras
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/deislabs/ratify/pkg/common"
+	"github.com/deislabs/ratify/pkg/ocispecs"
+	"github.com/deislabs/ratify/pkg/referrerstore"
+	"github.com/dgraph-io/ristretto"
+	"github.com/dgraph-io/ristretto/z"
+	"github.com/sirupsen/logrus"
+)
+
+// operation defines the API operations of ReferrerStore.
+type operation int
+
+const (
+	operationName operation = iota
+	operationGetConfig
+	operationListReferrers
+	operationGetBlobContent
+	operationGetReferenceManifest
+	operationGetSubjectDescriptor
+)
+
+type orasStoreWithInMemoryCache struct {
+	referrerstore.ReferrerStore
+	cache     *ristretto.Cache
+	cacheConf *CacheConf
+}
+
+type CacheConf struct {
+	Enabled bool `json:"cacheEnabled"`
+	Ttl     int  `json:"ttl"`
+}
+
+func keyToHash(key interface{}) (uint64, uint64) {
+	if key == nil {
+		return 0, 0
+	}
+	switch k := key.(type) {
+	case string:
+		return z.MemHashString(k), xxhash.Sum64String(k)
+	default:
+		return 0, 0
+	}
+}
+
+// createCachedStore creates a new oras store decorated with in-memory cache to cache
+// results of ListReferrers API.
+func createCachedStore(storeBase referrerstore.ReferrerStore, cacheConf *CacheConf) (referrerstore.ReferrerStore, error) {
+	memoryCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1e5,
+		MaxCost:     1 << 30,
```

Comment:
Curious, how did you decide in the cache size?

Reply:
`BufferItems` is the size of Get buffers which is recommended to set as 64. https://github.com/dgraph-io/ristretto#Config

But we do care about `NumCounters` and `MaxCost`. `NumCounters` is recommended to 10x the number of cache items. `MaxCost` depends on how we define the cost for each item. If the cost for each item is 1, then it's equivalent to total item number. But if the item cost relates to item size, then it's same as the total size.

I plan to make configs like `NumCounters` and `MaxCost` configurable once we decide the cache library.

Reply:
make sense. thanks for the clarification.